### DISCUSSION TITLE
Is it possible to sign in a user with a REST API call when using Credentials provider?

### DISCUSSION THREAD

**Discussion by maksonlee** [nextauthjs/next-auth#6011] [Help]
### Question 💬

We are creating an API-only application, means that there is no UI, if we use the Credentials provider, is it possible to sign in a user with a REST API call and then get a token back?

### How to reproduce ☕️

Create an API-only Next.js application and use NextAuth.js for authentication.

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by ericvrp** [comment]
Could someone post some pointers to a solution/tutorial/guide here please?

**Reply by yuedong111** [reply]
After logging in, you should call /auth/sessions—checking the response status code—to retrieve the session. If you need the token itself, you can place the token’s contents into the session within your callback.

**Comment by Dragate** [comment]
Yes, it is possible (as documented by [docs](https://next-auth.js.org/getting-started/rest-api)). On a recent project I had a similar requirement, where the next.js API serves both the authentication of the site itself and of a react native app. The basic idea was to check the session endpoint to validate whether the user is authenticated and have 2 functions that do the sign in/out. The functions look something like this:
```js
    const signIn = async (credentials) => {
        return fetch(`${basePath}/api/auth/csrf`)
            .then(r => r.json())
            .then(({ csrfToken }) =>
                fetch(`${basePath}/api/auth/callback/credentials`, {
                    method: "POST",
                    headers: { "Content-Type": "application/json" },
                    body: JSON.stringify({ ...credentials, csrfToken, json: "true" }),
                })
            )
    }

    const signOut = async () => {
        return fetch(`${basePath}/api/auth/csrf`)
            .then(r => r.json())
            .then(({ csrfToken }) =>
                fetch(`${basePath}/api/auth/signout`, {
                    method: "POST",
                    headers: { "Content-Type": "application/json" },
                    body: JSON.stringify({ csrfToken, json: "true" }),
                })
            )
    }
```

You may also take inspiration from https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/react/index.tsx#L212

**Reply by JWPapi** [reply]
Hey I’m trying to connect to `api/auth/callback/credentials`

But I only get this back `{"url":"http://localhost:3000/api/auth/signin?csrf=true"}` (wrong or correct credentials)

my credentials authorize function is not even running  (checking with `console.log`)

Could you show how your [...nextauth.js] looks like?

**Reply by Husain010** [reply]
Any update on this? I am also getting the same thing.

**Reply by sadekxD** [reply]
Any update guys?

**Reply by isque03** [reply]
Thanks @Dragate! 
The above snippet worked for me. I was previously getting the response {"url":"http://localhost:3000/api/auth/signin?csrf=true"} from POST api/auth/callback/<credential-provider-id>, but I was using content type 'Content-Type': 'application/x-www-form-urlencoded'.

Changing to application/json fixed it for me. 

You must keep the cookies returned from the initial call to /api/auth/crsf.  

This is from the next-auth docs:

> GET /api/auth/csrf
> Returns object containing CSRF token. In NextAuth.js, CSRF protection is present on all authentication routes. It uses the "double submit cookie method", which uses a signed HttpOnly, host-only cookie.
> 
> The CSRF token returned by this endpoint must be passed as form variable named csrfToken in all POST submissions to any API endpoint.

The url for the /api/auth/callback rest endpoint wasn't totally clear to me intitially. I wasn't sure if it was always /api/auth/callback/credentials. But I confirmed the url format is:

`/api/auth/callback/<provider-id>`

and it's also listed in

`/api/auth/providers`

Here is an example assuming the provider id is 'test'. It takes username / password as credentials. I hope this helps someone, I felt like I wasted way too much time getting this to work.

```
const fetch = require('node-fetch');

async function testCredentialsSignIn() {
  const baseUrl = 'http://localhost:3000'; // The base URL of your local development server

  // Step 1: Get the CSRF Token
  const csrfResponse = await fetch(`${baseUrl}/api/auth/csrf`);
  const csrfData = await csrfResponse.json();
  const csrfToken = csrfData.csrfToken;

  console.log('CSRF Token:', csrfToken);

  // Step 2: Extract cookies from the CSRF response
  const csrfCookies = csrfResponse.headers.raw()['set-cookie'];
  console.log('Initial Cookies:', csrfCookies);

  // Step 3: Send a POST request to /api/auth/callback/test with CSRF token and cookies
  const signInResponse = await fetch(`${baseUrl}/api/auth/callback/test`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Cookie': csrfCookies.join(';'), 
    },
    body: JSON.stringify({
      csrfToken: csrfToken,
      username: 'test_user@example.com',
      password: 'some_password',
      json: 'true',
    }),
  });

  // Check the status and response headers
  console.log('Response status:', signInResponse.status);
  const signInCookies = signInResponse.headers.raw()['set-cookie'];
  console.log('Sign-In Cookies:', signInCookies);

  // Step 4: Combine cookies from CSRF and sign-in responses
  const combinedCookies = [...csrfCookies, ...signInCookies].join('; ');

  // Step 5: Fetch the session information after logging in
  const sessionResponse = await fetch(`${baseUrl}/api/auth/session`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      'Cookie': combinedCookies, // Use combined cookies
    },
  });

  const sessionInfo = await sessionResponse.json();
  console.log('Session Info:', sessionInfo);

  // Step 6: Sign out
  const signOutResponse = await fetch(`${baseUrl}/api/auth/signout`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Cookie': combinedCookies,
    },
    body: JSON.stringify({
      csrfToken: csrfToken,
      json: 'true',
    }),
  });


  console.log('Sign-Out Response status:', signOutResponse.status);
  const signOutBody = await signOutResponse.json();
  console.log('Sign-Out Response:', signOutBody);
}

testCredentialsSignIn().catch(console.error);

```

**Comment by ericvrp** [comment]
Thank you @Dragate! That was very helpful. 

Just like the original poster I am working on an API-only application. In my case it's not a NextJs one but a more barebones NodeJs script. Because my code is not running in the browser I had some more hurdles to take. For anyone in the future attempting to do the same I will post a list of facts, opinions (my own traps), and though processes.

1. I find it strange to call this the next-auth REST API. I got used to REST API's returning (most often) json data. This REST API returns webpages by default.
2. adding json="true" works to make it do what I need it for. But this feature flag is not documented on the REST documentation page as far as I can see.
3. I forgot to add `headers: { "Content-Type": "application/json" },` to the POST requests
4. because my fetch commands are not running in the browser I think cookies are not handled automatically. I hope someone can point me to the fetch docs that explains this!
5. to fix point 4. I now use https://github.com/valeriangalliat/fetch-cookie

Happy sailing with this great library from now on! 👍

**Comment by cryptofreedom** [comment]
next-auth doc is awful

**Comment by JWPapi** [comment]
What I did was a I built an custom endpoint for 3rd party applications, using the same credentials from the database, but practically no next-auth

**Comment by vozmarkov** [comment]
I end up creating my own custom endpoint to post credentials. I validate against db and return  JWT token, and in middleware, I check if the req contains api/v1/.. and a bearer token, also jtw includs roles[ Int id's] to handle access permission.

App db/business logic is made into helper functions,  I use the logic functions in api routes where next-auth isn't used and in server actions where auth is handled by next-auth. So far, it works pretty well.

**Comment by MobinAskari** [comment]
No matter which solution I choose, I'm either getting text instead of JSON as a response, or it just doesn't work, any working solutions as of 2024?

**Reply by vozmarkov** [reply]
Are you testing with Postman? make sure your post header is set correctly 
Something like Content-type: application/json

 Also, check middleware if it's not getting blocked in there. 

Whats the response status code.

**Reply by MobinAskari** [reply]
> Are you testing with Postman? make sure your post header is set correctly Something like Content-type: application/json
> 
> Also, check middleware if it's not getting blocked in there.
> 
> Whats the response status code.

It's just a custom script in Node.js being run with `tsx`

This solution just returns text (either the 404 page or the login page, and even a custom next-auth error page) instead of JSON regardless of using `/api/auth/callback/credentials` or `/api/auth/signin/credentials`: https://github.com/nextauthjs/next-auth/discussions/6011#discussioncomment-6359246 

This other solution https://github.com/nextauthjs/next-auth/discussions/6011#discussioncomment-10246459, returns: 
```
CSRF Token: 1f19b3b1a18c6316792897087cbbc8eb9e02a2d400b67eb3b493ac9fbcd3991b
Initial Cookies: [
  'authjs.csrf-token=1f19b3b1a18c6316792897087cbbc8eb9e02a2d400b67eb3b493ac9fbcd3991b%7Ce41f6f115aa1a2e92af0d52bdd2c8d033a62bea131ecc4a99a66e76cc4020163; Path=/; HttpOnly; SameSite=Lax',
  'authjs.callback-url=http%3A%2F%2Flocalhost%3A3000; Path=/; HttpOnly; SameSite=Lax'
]
Response status: 200
Sign-In Cookies: []
Session Info: null
Sign-Out Response status: 200
```

even though the credentials are correct, the session info is null, also I had to change `signInResponse.headers.raw()['set-cookie']` to `signInResponse.headers.getSetCookie()` which returns an empty array, hence the issue
